### PR TITLE
Graphite: Migrate to React (part 3: migrate segments)

### PR DESCRIPTION
**Pull request by ifrost**
**What this PR does / why we need it**:

Follow-up on #37070 and plan described in #36481. This PR replaces angularjs' metric segment and dropdowns with React's Segment.

**Which issue(s) this PR fixes**:

Fixes N/A

**Special notes for your reviewer**:

* There are some changes in grafana-ui to allow reloading options while typing in SegmentAsync (@grafana/grafana-frontend-platform)
* Focusing on newly added segments has been removed. It's not supported by React's Segment and it was quite hidden feature as in the previous version there was no UI feedback that a segment was in focus
* Fixes a bug: after removing all tags it's not possible to add metrics again (see also: https://github.com/grafana/grafana/pull/36484#pullrequestreview-710406682)
* Manual test scenarios for migration are defined [here](https://gist.github.com/ifrost/e8faa2976caf610e9e03ec815ce83d82).

### CODE REVIEW

File: public/app/plugins/datasource/graphite/components/TagsSection.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      return await getTagsAsSegmentsSelectables(state, inputValue);
+    },
+    [state]
+  );
+  const debouncedGetTagsAsSegments = useMemo(() => debounce(getTagsAsSegmentsOptions, 200, { leading: true }), [
+    getTagsAsSegmentsOptions,
+  ]);
+
+  return (
+    <div style={{ display: 'flex', flexDirection: 'row' }}>
```
Comment (gabor) [inline public/app/plugins/datasource/graphite/components/TagsSection.tsx:46]: maybe move these to `getStyles`?

File: public/app/plugins/datasource/graphite/components/MetricSegment.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * Options for tag names and metric names are reloaded while user is typing with backend taking care of auto-complete
+ * (auto-complete cannot be implemented in front-end because backend returns only limited number of entries)
+ *
+ * getAltSegmentsSelectables() also returns list of tags for segment with index=0. Once a tag is selected the editor
+ * enters tag-adding mode (see SeriesSection and GraphiteQueryModel.seriesByTagUsed).
+ */
+export function MetricSegment({ dispatch, metricIndex, segment, state }: Props) {
+  const loadOptions = useCallback(
+    async (value: string | undefined) => {
+      return await getAltSegmentsSelectables(state, metricIndex, value || '');
```
Comment (gabor) [inline public/app/plugins/datasource/graphite/components/MetricSegment.tsx:31]: is `await` here necessary?

**Review by gabor — COMMENTED** [review]
i added some nitpick-ish comments about the code, will test the functionality and report those separately.

**Comment by gabor** [comment]
there's this behavior in the react version that is not there in the angular version:
- open the graphite query editor
- click on select metric, and choose "stats"
- when you choose "stats", the dropdown closes, but the name on teh segment stays "select metric" for a very short time, and then that name changes to "stats". this does not take long, but it makes the segment "blink". this does not happen with the angular version

**Review by gabor — APPROVED** [review]
code looks good (except some nitpicks), works well, LGTM!

**Comment by Elfo404** [comment]
There's a weird issue happening when the graphite is down making the metric segment disappear. It's easy to reproduce:

https://user-images.githubusercontent.com/1170767/128032254-b5d9c211-1196-4e66-9b01-b2e253eedc99.mov

Note that graphite is not running in this case

**Comment by Elfo404** [comment]
Not sure if related to this PR or not, but the background for the icon segment button looks off, it seems like it's using a lighter shade of background:

Screenshot 2021-08-03 at 16 27 08

**Comment by ifrost** [comment]
> There's a weird issue happening when the graphite is down making the metric segment disappear. It's easy to reproduce:
> 
>  Screen.Recording.2021-08-03.at.16.20.54.mov 
> Note that graphite is not running in this case

It's the same bug in main - it happens when a function is selected without a metric. I'll have a look if it's easy to fix, though.

**Comment by ifrost** [comment]
> > There's a weird issue happening when the graphite is down making the metric segment disappear. It's easy to reproduce:
> > Screen.Recording.2021-08-03.at.16.20.54.mov
> > Note that graphite is not running in this case
> 
> It's the same bug in main - it happens when a function is selected without a metric. I'll have a look if it's easy to fix, though.

Hmm... not sure if there's an easy way to fix it without changing too much. Some functions do not accept series list (e.g. constantLine). It seems like disappearing metric select allows adding such functions :/ It's exactly the same in angularjs version but I'll revisit it at the end.

File: public/app/plugins/datasource/graphite/components/TagEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+ * also used for removing tag (with a special "--remove tag--" option provided by getTagsSelectables).
+ *
+ * Options for tag names and values are reloaded while user is typing with backend taking care of auto-complete
+ * (auto-complete cannot be implemented in front-end because backend returns only limited number of entries)
+ */
+export function TagEditor({ dispatch, tag, tagIndex, state }: Props) {
+  const getTagsOptions = useCallback(
+    async (inputValue: string | undefined) => {
+      return await getTagsSelectables(state, tagIndex, inputValue || '');
+    },
```
Comment (Elfo404) [inline public/app/plugins/datasource/graphite/components/TagEditor.tsx:29]: is async/await necessary here?
```suggestion
    (inputValue: string | undefined) => {
      return getTagsSelectables(state, tagIndex, inputValue || '');
    },
```

File: public/app/plugins/datasource/graphite/components/TagEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    [state, tagIndex]
+  );
+  const debouncedGetTagsOptions = useMemo(() => debounce(getTagsOptions, 200, { leading: true }), [getTagsOptions]);
+
+  const getTagValueOptions = useCallback(
+    async (inputValue: string | undefined) => {
+      return await getTagValuesSelectables(state, tag, tagIndex, inputValue || '');
```
Comment (Elfo404) [inline public/app/plugins/datasource/graphite/components/TagEditor.tsx:36]: and here :) 
```suggestion
    (inputValue: string | undefined) => {
      return getTagValuesSelectables(state, tag, tagIndex, inputValue || '');
```

File: public/app/plugins/datasource/graphite/components/TagsSection.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  const styles = useStyles2(getStyles);
+
+  const newTagsOptions = mapSegmentsToSelectables(addTagSegments || []);
+
+  // Options are reloaded while user is typing with backend taking care of auto-complete (auto-complete cannot be
+  // implemented in front-end because backend returns only limited number of entries)
+  const getTagsAsSegmentsOptions = useCallback(
+    async (inputValue: string) => {
+      return await getTagsAsSegmentsSelectables(state, inputValue);
+    },
```
Comment (Elfo404) [inline public/app/plugins/datasource/graphite/components/TagsSection.tsx:38]: same here, is async/wait needed? 
```suggestion
  const getTagsAsSegmentsOptions = useCallback(
    (inputValue: string) => {
      return getTagsAsSegmentsSelectables(state, inputValue);
    },
```

File: public/app/plugins/datasource/graphite/specs/query_ctrl.test.ts [hunk 6]
```diff
@@ -200,19 +202,20 @@ describe('GraphiteQueryCtrl', () => {
   });
 
   describe('when autocomplete for metric names is not available', () => {
-    silenceConsoleOutput();
+    // silenceConsoleOutput();
     beforeEach(() => {
       ctx.ctrl.state.datasource.getTagsAutoComplete = jest.fn().mockReturnValue(Promise.resolve([]));
       ctx.ctrl.state.datasource.metricFindQuery = jest.fn().mockReturnValue(
         new Promise(() => {
+          console.log('throw new Error();');
```
Comment (Elfo404) [inline public/app/plugins/datasource/graphite/specs/query_ctrl.test.ts:210]: is this intentional or is it a leftover?

**Review by Elfo404 — APPROVED** [review]
Cool stuff! 🚀 

seems to work as expected, left a couple of comments but no blockers.
I (again) really like the approach of migrating the editor piece by piece, it makes reviewing so much easier!

File: public/app/plugins/datasource/graphite/components/TagsSection.tsx (hunk 1 above)
Comment (ifrost) [reply public/app/plugins/datasource/graphite/components/TagsSection.tsx:46]: - [ ] use getStyles()

File: public/app/plugins/datasource/graphite/components/MetricSegment.tsx (hunk 2 above)
Comment (ifrost) [reply public/app/plugins/datasource/graphite/components/MetricSegment.tsx:31]: - [ ] remove redundant awaits

File: public/app/plugins/datasource/graphite/components/TagEditor.tsx (hunk 3 above)
Comment (ifrost) [reply public/app/plugins/datasource/graphite/components/TagEditor.tsx:29]: - [ ] remove redundant async/await

File: public/app/plugins/datasource/graphite/components/TagEditor.tsx (hunk 4 above)
Comment (ifrost) [reply public/app/plugins/datasource/graphite/components/TagEditor.tsx:36]: - [ ] remove redundant async/await

File: public/app/plugins/datasource/graphite/components/TagsSection.tsx (hunk 5 above)
Comment (ifrost) [reply public/app/plugins/datasource/graphite/components/TagsSection.tsx:38]: - [ ] remove redundant async/await

File: public/app/plugins/datasource/graphite/specs/query_ctrl.test.ts (hunk 6 above)
Comment (ifrost) [reply public/app/plugins/datasource/graphite/specs/query_ctrl.test.ts:210]: - [ ] revert accidental debug changes :)

File: public/app/plugins/datasource/graphite/components/TagsSection.tsx (hunk 1 above)
Comment (ifrost) [reply public/app/plugins/datasource/graphite/components/TagsSection.tsx]: f331c847f558cc97feb7af774c4e49b1c5b347e5

**Comment by ifrost** [comment]
> there's this behavior in the react version that is not there in the angular version:
> 
> * open the graphite query editor
> * click on select metric, and choose "stats"
> * when you choose "stats", the dropdown closes, but the name on teh segment stays "select metric" for a very short time, and then that name changes to "stats". this does not take long, but it makes the segment "blink". this does not happen with the angular version

Should be fixed by ae24255b147260cdfe79a3b91df323c4b7c28dac. It is the default behavior of react-select to always display the selected value in the input, but... react-select expects objects passed as `{value: string, label: string}`. Our wrappers (`SegmentAsync` asn `Segment`) allow to pass such objects or plain strings. When a plain string is passed as a value then it's mapped to react-select object but without `label` property ([here](https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/Segment/Segment.tsx#L66)) effectively changing the behavior of react-select to not show the label. I've changed MetricSegment so the string is passed instead of the object to get rid of that text. I think it looks better.

File: packages/grafana-ui/src/components/Segment/SegmentAsync.tsx [hunk 7]
```diff
@@ -13,6 +13,9 @@ import { useStyles } from '../../themes';
 export interface SegmentAsyncProps<T> extends SegmentProps<T>, Omit<HTMLProps<HTMLDivElement>, 'value' | 'onChange'> {
   value?: T | SelectableValue<T>;
   loadOptions: (query?: string) => Promise<Array<SelectableValue<T>>>;
+  // if true, options will be reloaded when user changes the value in the input,
```
[the remark this answers is not in the document]
Comment (ifrost) [reply packages/grafana-ui/src/components/Segment/SegmentAsync.tsx:16]: - [ ] use block docs

**Review by tskarhed — APPROVED** [review]

**Pull request merged**